### DISCUSSION TITLE
How to work with database/cache, eg Hive?

### DISCUSSION THREAD

**Discussion by dreamer2q** [rrousselGit/riverpod#1261] [Q&A]
A common use-case is that, an App can cache the HTTP request they make. When the next time the App launches, it first tries to load data from cache. If the cache is hit, then display it from cache. Otherwise, when the cache is missed,  just make a HTTP request to fetch latest data and cache it. Users can pull-down to refresh, which means sending a new HTTP request and update the cache. If the updating process goes wrong, users can see an error and stay with the cached data.

I chose Hive as the cache service. But when it comes to implementing, I found it quite hard to work with riverpod features.

e.g, I want to use the feature `ref.refresh` to do refreshing. But it turned out I can hardly control the data flow. Because, I cannot determine whether the `refresh` is first created or recreated. In order to do so, I need a flag telling me how the `refresh` is invoked.

**Comment by hamza-imran75** [comment]
Andrea Biz has repository about data persistent maybe this is what you are looking for.

https://github.com/bizz84/movie_app_state_management_flutter

**Comment by mark8044** [comment]
I just integrated my SQLite database with riverpod and its working flawlessly.  Heres what I did, maybe it can help you:

```
final DBProvider = ChangeNotifierProvider<DBData>((ref) {
  return DBData();
});
```

```
class DBData with ChangeNotifier
{

  var _theDB;
  List _repliesResults = [];

  List get repliesResults => _repliesResults;


  _init() async {
    _theDB = await openDatabase(
      // Set the path to the database. Note: Using the `join` function from the
      // `path` package is best practice to ensure the path is correctly
      // constructed for each platform.
      join(await getDatabasesPath(), 'my_database.db'),
    );

  }

  void refreshAll() async {
    getReply();
  }

  void deleteAll() async {
    await _init();

    // Remove from the database.
    _theDB.rawDelete("Delete from replies");

    refreshAll();
  }

  void getReply() async {
    // Query the table
    await _init();

    final List<Map<String, dynamic>> maps = await _theDB.query('my_replies');

    _repliesResults = List.generate(maps.length, (i) {
      return maps[i];
    });

    notifyListeners();
  }
}
```

This `ChangeNotifier` actually has a lot more functions to it that do various things, but I tried to simply it to post here

Then I can use the results as you may do with other ChangeNotifier

For example, in a listview.builder I can get a realtime update of what's contained in my DB:
```
             return ListView.builder(
                              itemCount: ref.watch(DBProvider).repliesResults.length,
                              itemBuilder: (context, index) {
                                           return MyReplyItem(data:ref.watch(DBProvider).repliesResults[index]);
                            })

```

The first DB data pull when the screen is opened is done in the initState
```
  @override
  void initState() {
    super.initState();
    ref.read(DBProvider).refreshAll();
  }
```

Or if I need to explicitly `deleteAll` from my database:

```
ref.read(DBProvider).deleteAll();
```


Hope this helps in some way, it was quite easy and effective to implement IMO and works super fast with real time DB updates
